src/lib/homestore_backend/replication_state_machine.cpp

```diff
@@ -478,6 +480,163 @@ void ReplicationStateMachine::set_snapshot_context(std::shared_ptr< homestore::s
     m_snapshot_context = context;
 }
 
+folly::Future< std::error_code > ReplicationStateMachine::on_fetch_data(const int64_t lsn, const sisl::blob& header,
+                                                                        const homestore::MultiBlkId& local_blk_id,
+                                                                        sisl::sg_list& sgs) {
+    // the lsn here will always be -1 ,since this lsn has not been appeneded and thus get no lsn
+
+    const ReplicationMessageHeader* msg_header = r_cast< const ReplicationMessageHeader* >(header.cbytes());
+
+    if (msg_header->corrupted()) {
+        LOGW("replication message header is corrupted with crc error, lsn:{}", lsn);
+        return folly::makeFuture< std::error_code >(std::make_error_code(std::errc::bad_message));
+    }
+
+    LOGD("fetch data with lsn: {}, msg type: {}", lsn, msg_header->msg_type);
+
+    // for nuobject case, we can make this assumption, since we use append_blk_allocator.
+    RELEASE_ASSERT(sgs.iovs.size() == 1, "sgs iovs size should be 1, lsn: {}, msg_type: {}", lsn, msg_header->msg_type);
+
+    auto const total_size = local_blk_id.blk_count() * repl_dev()->get_blk_size();
+    RELEASE_ASSERT(total_size == sgs.size,
+                   "total_blk_size does not match, lsn: {}, msg_type: {}, expected size {}, given buffer size {}", lsn,
+                   msg_header->msg_type, total_size, sgs.size);
+
+    auto given_buffer = (uint8_t*)(sgs.iovs[0].iov_base);
+
+    switch (msg_header->msg_type) {
+    case ReplicationMessageType::CREATE_SHARD_MSG:
+    case ReplicationMessageType::SEAL_SHARD_MSG: {
+        // this function only returns data, not care about raft related logic, so no need to check the existence of
+        // shard, just return the shard header/footer directly. Also, no need to read the data from disk, generate it
+        // from Header.
+        auto sb =
+            r_cast< HSHomeObject::shard_info_superblk const* >(header.cbytes() + sizeof(ReplicationMessageHeader));
+
+        auto const expected_size =
+            sisl::round_up(sizeof(HSHomeObject::shard_info_superblk), repl_dev()->get_blk_size());
+
+        RELEASE_ASSERT(
+            sgs.size == expected_size,
+            "shard metadata size does not match, lsn: {}, msg_type: {}, expected size {}, given buffer size {}", lsn,
+            msg_header->msg_type, expected_size, sgs.size);
+
+        // TODO：：return error_code if assert fails, so it will not crash here because of the assert failure.
+
+        std::memcpy(given_buffer, sb, total_size);
+        return folly::makeFuture< std::error_code >(std::error_code{});
+    }
+
+        // TODO: for shard header and footer, follower can generate it itself according to header, no need to fetch it
+        // from leader. this can been done by adding another callback, which will be called before follower tries to
+        // fetch data.
+
+    case ReplicationMessageType::PUT_BLOB_MSG: {
```

Comment:
I probably didn't quite get fetch blob would have message type put_blob_msg? Is there a reason behind it?

Reply:
user header is included in the fetch_data request https://github.com/eBay/HomeStore/blob/36600b1ae949347f0ce652ee7f552c07e298bedd/src/lib/replication/fetch_data_rpc.fbs#L7
so , at homeobject level, we can identify the message type from the user header.
actually , the message type here means which kind of  raft op we are trying to fetch data for.  PUT_BLOB_MSG means we received a put_blob in leader, and the corresponding log has been sent to and received by follower, but follower does not receive the data, so this follower wants to get the data of this PUT_BLOB_MSG message( which is also the type of the corresponding log ).